K8s Proxy Attempt 3

liamrathke opened the pull request:
This will be edited to line up with Backstage PR standards

jamieklassen commented on plugins/kubernetes-backend/src/service/KubernetesProxy.ts:
```diff
@@ -0,0 +1,289 @@
+      };
+    }
+
+    const ca = details.caData;
+    if (!ca) {
+      this.logger.error('could not find certificate!');
+      return {
+        code: ERROR_INTERNAL_SERVER,
+        data: {
+          error: 'Invalid CA certificate configured within Backstage',
+        },
+      };
+    }
```
Sometimes a `ClusterDetails` will not have `caData`, especially if [`skipTLSVerify`](https://github.com/backstage/backstage/blob/c7dd03bf96daa7e4d02fba82137d9a5b41e8b400/plugins/kubernetes-backend/src/types/types.ts#L157) is true. Can you make this code resilient to those cases?

jamieklassen commented on plugins/kubernetes-backend/src/service/KubernetesProxy.ts:
```diff
@@ -0,0 +1,289 @@
+      };
+    }
+
+    const agent = new https.Agent({ ca });
```
the `https.Agent` constructor's parameters wind up getting [passed](https://github.com/nodejs/node/blob/a3e110820ff98702e1761831e7beaf0f5f1f75e7/lib/https.js#L1500) to `tls.connect`, which ultimately [passes](https://github.com/nodejs/node/blob/9cde7a033e66a8e9e691c708cfb448161fa2c27d/lib/internal/tls/secure-context.js#L62) the `ca` parameter to the C++ function `SecureContext::AddCACert`, where it gets [parsed](https://github.com/nodejs/node/blob/5fad0b93667ffc6e4def52996b9529ac99b26319/src/crypto/crypto_context.cc#L702-L703) by the OpenSSL function `PEM_read_bio_X509_AUX`, which according to [the docs](https://www.openssl.org/docs/man1.1.1/man3/PEM_read_bio_X509_AUX.html)
> read(s) ... structures in PEM format. In this sense PEM format is simply base64 encoded data surrounded by header lines

However, when backstage [sets up a KubeConfig](https://github.com/backstage/backstage/blob/a370706a18a3b8e3db8822b231a0bb554c8faf6c/plugins/kubernetes-backend/src/service/KubernetesClientProvider.ts#L49-L54) object from `@kubernetes/client-node` normally, that object will [set HTTPS options](https://github.com/kubernetes-client/javascript/blob/6131baeb1ca46337597b458a0b73ccc0261cfd9b/src/config.ts#L406-L409) on k8s API requests by first [base64-decoding](https://github.com/kubernetes-client/javascript/blob/6131baeb1ca46337597b458a0b73ccc0261cfd9b/src/config.ts#L513-L515) the value from `caData`.

All that to say, can you make sure to base64-decode the `caData` value here rather than using it verbatim?

jamieklassen replied:
Also I'll be the first to admit that the [docs on this topic](https://backstage.io/docs/features/kubernetes/configuration#clusterscadata-optional), which say

> PEM-encoded certificate authority certificates

feel a bit misleading for not mentioning the extra level of base64-encoding that is expected (even though that is subtly implied by the "inspecting the Kubernetes config file" instruction)

jamieklassen replied:
I took a [stab](https://github.com/backstage/backstage/pull/12911) at fixing those docs

mklanjsek commented on plugins/kubernetes-backend/src/kubernetes-auth-translator/TokenKubernetesAuthTranslator.ts:
```diff
@@ -0,0 +1,39 @@
+import { ClusterDetails } from '../types/types';
+import { KubernetesRequestAuth } from '@backstage/plugin-kubernetes-common';
+
+export class TokenKubernetesAuthTranslator implements KubernetesAuthTranslator {
```
Shouldn't that just be something like this:
```    
if (!authToken) {
   throw new Error('Token not found under auth.token in request body');
 }

return {
  ...clusterDetails,
  serviceAccountToken: authToken,
}
```
That's easier to read and should do the same thing.

liamrathke replied:
Agreed that this format is better - however, I copied this file from the [`OidcKubernetesAuthTranslator`](https://github.com/backstage/backstage/blob/master/plugins/kubernetes-backend/src/kubernetes-auth-translator/OidcKubernetesAuthTranslator.ts), and decided at the time that it would be better to keep all auth translators as similar as possible, for consistency. 

Not sure if it would make sense to change this auth translator but keep the others the same, or change the other auth translators to match this new format as part of this specific PR.

mklanjsek commented on plugins/kubernetes-backend/src/service/KubernetesBuilder.ts:
```diff
@@ -325,4 +337,29 @@ export class KubernetesBuilder {
+      this.buildClusterSupplier(this.defaultClusterRefreshInterval);
+
+    if (!kcs) {
+      // error
```
nit: remove comment

mklanjsek reviewed (changes requested):
Looks really good!  Let me know when you add tests - I would like to take another look then.

jamieklassen commented on plugins/kubernetes-backend/src/service/KubernetesProxy.ts:
```diff
@@ -0,0 +1,290 @@
+  ): Promise<ClusterDetails[]> {
+    const clusters = await clusterSupplier.getClusters();
+
+    const clusterNames = Object.keys(krc);
+
+    const clusterDetails = clusters.filter(c => clusterNames.includes(c.name));
+
+    const clusterDetailsAuth = await Promise.all(
+      clusterDetails.map(async c => {
+        const authMetadata: KubernetesRequestAuth = {
+          token: krc[c.name],
+        };
+        const cAuth = await this.configureAuth(c, authMetadata);
+        return cAuth;
+      }),
+    );
```
If we're going to take @chinigo's [advice about auth](https://github.com/backstage/backstage/issues/12231#issuecomment-1192875578):

> The proxy would expect calls to supply an opaque `Authorization` header, and it will blindly send them along in its fanned-out requests. Operators who want to support multi-cluster operations can use their platforms to respect that shared token at their discretion

I think there may be no need to introduce the `TokenKubernetesAuthTranslator` -- instead of using a KubernetesAuthTranslator at all, this code could enrich the clusterDetails with the token directly, like this (maybe with some copying instead of mutation in the `map`, i'm rusty):
```suggestion
    const clusterNames = Object.keys(krc);

    const clusterDetails = clusters.filter(c => clusterNames.includes(c.name));

    const clusterDetailsAuth = await Promise.all(
      clusterDetails.map(async c => {
        c.serviceAccountToken = krc[c.name];
        return c;
      }),
    );
```

jamieklassen replied:
I might have gone a bit far here -- I think there could still be a place for a `configureAuth` method, but it probably doesn't need to invoke any KubernetesAuthTranslators in this version.

liamrathke replied:
I originally added this logic with the auth translator so that we could fallback if the token isn't included in the header; later on, if there isn't a token, we can use the generic auth translator for the cluster to attempt to authenticate.

Not sure if we need the fallback functionality any more, but if we want to keep it, then I would lean towards going through the auth translator for this.

jamieklassen replied:
I believe that @chinigo's "opaque header" idea requires there to be _no_ fallback functionality, as it represents a pure security risk until we've thought carefully about other ways of securing this endpoint.

Not to say that we don't want this "fallback"-type convenience eventually, but it feels like it could become a can of worms and I don't want your PR to get bogged down!

jamieklassen commented:
Something that could go a long way here is giving a decent indication of how frontend code could consume this endpoint. I'm imagining extending the [`KubernetesApi` interface](https://github.com/backstage/backstage/blob/44b2669c9bd6e0165bc2271da9ef12466c9c01ec/plugins/kubernetes/src/api/types.ts#L27-L38) to include a method for getting an arbitrary kubernetes API client, something like
```diff
diff --git a/plugins/kubernetes/src/api/types.ts b/plugins/kubernetes/src/api/types.ts
index 7c5943a6eb..491387fc62 100644
--- a/plugins/kubernetes/src/api/types.ts
+++ b/plugins/kubernetes/src/api/types.ts
@@ -19,6 +19,7 @@ import {
   ObjectsByEntityResponse,
 } from '@backstage/plugin-kubernetes-common';
 import { createApiRef } from '@backstage/core-plugin-api';
+import type { ApiConstructor, ApiType } from '@kubernetes/client-node';

 export const kubernetesApiRef = createApiRef<KubernetesApi>({
   id: 'plugin.kubernetes.service',
@@ -35,4 +36,8 @@ export interface KubernetesApi {
       oidcTokenProvider?: string | undefined;
     }[]
   >;
+  proxyApiClient<T extends ApiType>(
+    clusters: { [key: string]: string },
+    apiClientType: ApiConstructor<T>,
+  ): T;
 }
```

which could then be called like

```typescript
import { CustomObjectsApi } from '@kubernetes/client-node';

const kubernetesApi = useApi(kubernetesApiRef);
const clusterTokenMap = {}; // figure out how to get tokens for each cluster of interest
const customObjects = kubernetesApi.proxyApiClient(
  clusterTokenMap,
  CustomObjectsApi,
);
const response = await customObjects.listClusterCustomObject(
  'kappctrl.k14s.io',
  'v1alpha1',
  'apps',
);
```

Even better (but probably not required for the PR to be mergeable) would be to have a sample plugin which does this in the contrib folder.

jamieklassen commented on plugins/kubernetes-backend/src/service/KubernetesBuilder.ts:
```diff
@@ -273,6 +282,9 @@ export class KubernetesBuilder {
       });
     });
 
+    router.get('/proxy/:encodedQuery', this.makeProxyRequest.bind(this));
+    router.post('/proxy/:encodedQuery', this.makeProxyRequest.bind(this));
```
we're only supporting GET and POST here, but the kubernetes API [supports](https://kubernetes.io/docs/reference/using-api/api-concepts/#api-verbs) the full range of HTTP verbs. It's OK to narrow our focus to a few for a first offering, but we should make it explicit (say, in documentation) if we do.

jamieklassen commented:
There's a bit of a JSON bias here -- some core k8s APIs (like logs, port-forward and exec obviously) do not speak JSON so this implementation can't support those, and I'm pretty confident that clusters can contain [API aggregation layer](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/apiserver-aggregation/) services which I think are super flexible and not necessarily required to speak JSON.

jamieklassen commented on plugins/kubernetes-backend/src/service/KubernetesProxy.ts:
```diff
@@ -0,0 +1,290 @@
+    try {
+      const req = await fetch(uri, reqData);
+
+      const res = await req.json();
```
There's a bit of a JSON bias here -- some core k8s APIs (like logs, port-forward and exec obviously) do not speak JSON so this implementation can't support those, and I'm pretty confident that clusters can contain [API aggregation layer](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/apiserver-aggregation/) services which are not necessarily required to speak JSON.

jamieklassen commented on plugins/kubernetes-backend/src/service/KubernetesProxy.ts:
```diff
@@ -0,0 +1,290 @@
+    return clusterDetailsAuth;
+  }
+
+  protected getClusterIP(details: ClusterDetails): string {
```
FWIW the `server` field on a cluster is a URI, not an IP -- it will have a scheme like `https://` or and it can include a DNS hostname rather an an IP (EKS clusters look like `https://IDENTIFIER.STUFF.REGION.eks.amazonaws.com`)

liamrathke committed:
feat: Liam Rathke pre RFC update commit squash

Signed-off-by: Liam Rathke <liam.rathke@gmail.com>

Copy of proxy plugin work

Signed-off-by: Liam Rathke <liam.rathke@gmail.com>

Adds B64 encoding, TLS verify check

Signed-off-by: Liam Rathke <liam.rathke@gmail.com>

Sets up test scaffolding

Signed-off-by: Liam Rathke <liam.rathke@gmail.com>

Finishes proxy implementation tests

Signed-off-by: Liam Rathke <liam.rathke@gmail.com>

Removes deprecated buffer method

Signed-off-by: Liam Rathke <liam.rathke@gmail.com>

Adds additional HTTP verbs

Signed-off-by: Liam Rathke <liam.rathke@gmail.com>

Removes fallback

Signed-off-by: Liam Rathke <liam.rathke@gmail.com>

Adds some content-agnosticness, removes nitpick

Signed-off-by: Liam Rathke <liam.rathke@gmail.com>

Adds changeset

Signed-off-by: Liam Rathke <liam.rathke@gmail.com>

Fixes TSC errors?

Signed-off-by: Liam Rathke <liam.rathke@gmail.com>

Adds API Report docs

Signed-off-by: Liam Rathke <liam.rathke@gmail.com>

Removes unnecessary KubernetesRequestAuth parameter

Signed-off-by: Liam Rathke <liam.rathke@gmail.com>

Somehow adds new docs???

Signed-off-by: Liam Rathke <liam.rathke@gmail.com>

Fixes some code inline with review comments

Signed-off-by: Liam Rathke <liam.rathke@gmail.com>

Updates API report

Signed-off-by: Liam Rathke <liam.rathke@gmail.com>

Fixes tests

Signed-off-by: Liam Rathke <liam.rathke@gmail.com>

freben committed:
added a by-refs batch endpoint for entities

Signed-off-by: Fredrik Adelöw <freben@gmail.com>

awanlin committed:
Added SidebarSubmenu for all Kinds

Signed-off-by: Andre Wanlin <67169551+awanlin@users.noreply.github.com>

Commit:
feat: Carlos Lopez pre RFC update commit squash

docs: Remove unnecessary types
test: Use MSW instead of mocking fetch
fix: Updated api-report.md & error stringify from @backstage/errors
feat: Kubernetes builder getters/setters update
fix: Error handling with fix: Error handling with`@backstage/errors`
fix: Use same MSW version as other plugins
fix: Sort handleProxyRequest inputs
fix: Small type & comment fixes
fix: Change serverIP to serverURI
fix: Variable name improvements

Signed-off-by: Carlos Esteban Lopez <lcarlosesteb@vmware.com>

Commit:
feat: Remove multi-cluster support per latest RFC version

Signed-off-by: Carlos Esteban Lopez <lcarlosesteb@vmware.com>

Commit:
fix: Remove unused type in @backstage/kubernetes-common

Signed-off-by: Carlos Esteban Lopez <lcarlosesteb@vmware.com>

Commit:
fix: Handle skipTLSVerify properly

Signed-off-by: Carlos Esteban Lopez <lcarlosesteb@vmware.com>

Commit:
fix: Update test urls to https

Signed-off-by: Carlos Esteban Lopez <lcarlosesteb@vmware.com>

awanlin committed:
Changes based on feedback

Signed-off-by: Andre Wanlin <67169551+awanlin@users.noreply.github.com>

muhammadnmuhammad01 committed:
yarn changeset

Signed-off-by: Mohamed Mohamed <mohamed.mohamed2@aa.com>

muhammadnmuhammad01 committed:
pr review changes

Signed-off-by: Mohamed Mohamed <mohamed.mohamed2@aa.com>

freben committed:
implement by-refs batch fetching in the catalog client

Signed-off-by: Fredrik Adelöw <freben@gmail.com>

dagda1 committed:
improve the Custom ScmAuthApi Implementation section of docs/auth

Signed-off-by: Paul Cowan <paul.cowan@cutting.scot>

dagda1 committed:
run prettier

Signed-off-by: Paul Cowan <paul.cowan@cutting.scot>

diegoh committed:
Adding FanDuel to adopters

Signed-off-by: Diego Herrera <608492+diegoh@users.noreply.github.com>

Znarvl committed:
ran prettier, again

Signed-off-by: Simon <simme.jakobsson@gmail.com>

Rugvip committed:
Merge pull request #14632 from diegoh/adding-fanduel-to-adopters

Adding FanDuel to adopters

Rugvip committed:
Merge pull request #14627 from backstage/rugvip/build

create-app: switch root build script

Znarvl committed:
updated api-report

Signed-off-by: Simon <simme.jakobsson@gmail.com>

graynk committed:
deprecate UserSettingsTab in favour of SettingsLayout.Route

Signed-off-by: Nikita Karpukhin <nikita.karpukhin@scout24.com>

graynk committed:
combine two changesets

Signed-off-by: Nikita Karpukhin <nikita.karpukhin@scout24.com>

Rugvip committed:
Merge pull request #14631 from backstage/rugvip/polish

changsets: tweak changeset to make vale happy an elaborate a bit more

github-actions[bot] committed:
Version Packages

Commit:
Cost Insights Configurable Default Currency

Signed-off-by: bnechyporenko <bnechyporenko@bol.com>

acierto committed:
Generated Api Report

Signed-off-by: bogdannechyporenko <bogdannechiporenko@gmail.com>

acierto committed:
Added a changeset

Signed-off-by: bogdannechyporenko <bogdannechiporenko@gmail.com>

freben committed:
verify at CI that the lockfile has no duplicates

Signed-off-by: Fredrik Adelöw <freben@gmail.com>

freben committed:
Merge pull request #14521 from backstage/changeset-release/master

Version Packages

Commit:
microsite: Add blog entry about LF course

Signed-off-by: Jorge Lainfiesta <jorge@roadie.io>

freben committed:
Merge pull request #14630 from backstage/freben/dedupe

run yarn dedupe

acierto committed:
Added test cases

Signed-off-by: bogdannechyporenko <bogdannechiporenko@gmail.com>

acierto committed:
Updated the documentation of cost insight plugin

Signed-off-by: bogdannechyporenko <bogdannechiporenko@gmail.com>

acierto committed:
Fixed the typo

Signed-off-by: bogdannechyporenko <bogdannechiporenko@gmail.com>

freben committed:
add release notes for v1.8.0

Signed-off-by: Fredrik Adelöw <freben@gmail.com>

jhaals committed:
renovate: dedupe dependency bumps

Signed-off-by: Johan Haals <johan.haals@gmail.com>

Commit:
Add blog entry about LF course

Signed-off-by: Jorge Lainfiesta <jorge@roadie.io>

freben committed:
Merge pull request #14635 from backstage/freben/dupe

verify at CI that the lockfile has no duplicates

Commit:
Add truncate tag to blog post

Signed-off-by: Jorge Lainfiesta <jorge@roadie.io>

acierto committed:
Prettified the code

Signed-off-by: bogdannechyporenko <bogdannechiporenko@gmail.com>

freben committed:
Merge pull request #14618 from backstage/freben/catalog-client-tests

update tests to be more robust against catalog-client changes

jhaals committed:
npm -> yarn

Signed-off-by: Johan Haals <johan.haals@gmail.com>

priitliivak committed:
Update running-backstage-locally.md

https url is more accessible for environments, where github authentication has not been set up.

Signed-off-by: Priit Liivak <priitliivak@gmail.com>

emmaindal committed:
remove search alpha label in header

Signed-off-by: Emma Indal <emma.indahl@gmail.com>

jhaals committed:
Merge pull request #14637 from backstage/jhaals/rendedup

renovate: dedupe dependency bumps

jhaals committed:
Merge pull request #14640 from priitliivak/patch-1

Update running-backstage-locally.md

graynk committed:
Deprecate UserSettingsTab instead of removing it, add tests to check for tab content

Signed-off-by: Nikita Karpukhin <nikita.karpukhin@scout24.com>

freben committed:
Merge pull request #14638 from backstage/freben/v180

add release notes for v1.8.0

Znarvl committed:
updated changeset to conclude all updates in one file, changed description

Signed-off-by: Simon <simme.jakobsson@gmail.com>

luchillo17 committed:
Update plugins/kubernetes-backend/src/service/KubernetesProxy.ts

Co-authored-by: Jamie Klassen <jklassen@vmware.com>
Signed-off-by: Carlos Esteban Lopez Jaramillo <luchillo17@gmail.com>

luchillo17 committed:
Update plugins/kubernetes-backend/src/service/KubernetesProxy.ts

Co-authored-by: Jamie Klassen <jklassen@vmware.com>
Signed-off-by: Carlos Esteban Lopez Jaramillo <luchillo17@gmail.com>

luchillo17 committed:
Update plugins/kubernetes-backend/src/service/KubernetesProxy.ts

Co-authored-by: Jamie Klassen <jklassen@vmware.com>
Signed-off-by: Carlos Esteban Lopez Jaramillo <luchillo17@gmail.com>

luchillo17 committed:
Update plugins/kubernetes-backend/src/service/KubernetesProxy.ts

Co-authored-by: Jamie Klassen <jklassen@vmware.com>
Signed-off-by: Carlos Esteban Lopez Jaramillo <luchillo17@gmail.com>

luchillo17 committed:
Update plugins/kubernetes-backend/src/service/KubernetesProxy.ts

Co-authored-by: Jamie Klassen <jklassen@vmware.com>
Signed-off-by: Carlos Esteban Lopez Jaramillo <luchillo17@gmail.com>

dependabot[bot] committed:
build(deps): bump loader-utils from 1.4.1 to 1.4.2

Bumps [loader-utils](https://github.com/webpack/loader-utils) from 1.4.1 to 1.4.2.
- [Release notes](https://github.com/webpack/loader-utils/releases)
- [Changelog](https://github.com/webpack/loader-utils/blob/v1.4.2/CHANGELOG.md)
- [Commits](https://github.com/webpack/loader-utils/compare/v1.4.1...v1.4.2)

---
updated-dependencies:
- dependency-name: loader-utils
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>

andrewthauer committed:
fix: incorrect export type of GithubOrgEntityProvider

Signed-off-by: Andrew Thauer <athauer@wealthsimple.com>

jhaals committed:
document, add tests

Signed-off-by: Johan Haals <johan.haals@gmail.com>

freben committed:
add some eslint rules for testing-library use in tests

Signed-off-by: Fredrik Adelöw <freben@gmail.com>

Rugvip committed:
Merge pull request #14648 from andrewthauer/fix-gh-org-provider

fix: incorrect export type of GithubOrgEntityProvider

Rugvip committed:
value: add "sqs"

Signed-off-by: Patrik Oldsberg <poldsberg@gmail.com>

Rugvip committed:
Merge pull request #14650 from backstage/rugvip/works

vale: add "sqs"

Rugvip committed:
scripts/patch-release-for-pr: skip commit verification

Signed-off-by: Patrik Oldsberg <poldsberg@gmail.com>

Rugvip committed:
Merge pull request #14592 from backstage/rugvip/nolink

Revert "chore: (eslint) restrict imports of @material-ui Link"

Rugvip committed:
Merge pull request #14653 from backstage/rugvip/verify

scripts/patch-release-for-pr: skip commit verification

freben committed:
Update versioning policy to treat utility/service API additions as non-breaking

Co-authored-by: Patrik Oldsberg <poldsberg@gmail.com>
Co-authored-by: Johan Haals <johan.haals@gmail.com>
Signed-off-by: Fredrik Adelöw <freben@gmail.com>

Commit:
Update livestream URL on live page

Signed-off-by: Suzanne Daniels <suzanned@spotify.com>

freben committed:
Merge pull request #14658 from backstage/comsess-nov

Update livestream URL on live page

dagda1 committed:
nitialize all formData in next/Stepper

Signed-off-by: Paul Cowan <paul.cowan@cutting.scot>

freben committed:
review updates

Signed-off-by: Fredrik Adelöw <freben@gmail.com>

punkle committed:
wait for initialize to complete before finalize

Signed-off-by: Brian Fletcher <brian@roadie.io>

dagda1 committed:
give the ability to supply a transformErrors function to the Stepper form

Signed-off-by: Paul Cowan <paul.cowan@cutting.scot>

Znarvl committed:
fixed dot

Signed-off-by: Simon <simme.jakobsson@gmail.com>

freben committed:
Merge pull request #14583 from AmericanAirlines/update-create-plugin-template

feat: update create plugin template

dagda1 committed:
rim api-reports

Signed-off-by: Paul Cowan <paul.cowan@cutting.scot>

jhaals committed:
Merge pull request #14646 from backstage/dependabot/npm_and_yarn/loader-utils-1.4.2

build(deps): bump loader-utils from 1.4.1 to 1.4.2

dependabot[bot] committed:
build(deps): bump loader-utils from 1.4.1 to 1.4.2 in /storybook

Bumps [loader-utils](https://github.com/webpack/loader-utils) from 1.4.1 to 1.4.2.
- [Release notes](https://github.com/webpack/loader-utils/releases)
- [Changelog](https://github.com/webpack/loader-utils/blob/v1.4.2/CHANGELOG.md)
- [Commits](https://github.com/webpack/loader-utils/compare/v1.4.1...v1.4.2)

---
updated-dependencies:
- dependency-name: loader-utils
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>

dagda1 committed:
remove onChange handler

Signed-off-by: Paul Cowan <paul.cowan@cutting.scot>

Commit:
Adds surname to accepted vocab

Signed-off-by: Jorge Lainfiesta <jorge@roadie.io>

Rugvip committed:
Merge pull request #14396 from graynk/customize-user-settings-page

Make UserSettings page fully customizable

acierto committed:
Incorporated the feedback

Signed-off-by: bogdannechyporenko <bogdannechiporenko@gmail.com>

acierto committed:
Merge remote-tracking branch 'origin/master' into cost-insight-configurable-default-currency

acierto committed:
Merge remote-tracking branch 'origin/master' into scaffolder-result-content

Znarvl committed:
removed getRawBody with buffer, changed back depricatedTypes in config index and removed integration test

Signed-off-by: Simon <simme.jakobsson@gmail.com>

Znarvl committed:
fixt tsc error

Signed-off-by: Simon <simme.jakobsson@gmail.com>

Znarvl committed:
updated api-report

Signed-off-by: Simon <simme.jakobsson@gmail.com>

jhaals committed:
Merge pull request #14636 from jorgelainfiesta/add-blog-lf-course

microsite: Add blog entry about LF course

jhaals committed:
Merge pull request #14008 from kuangp/feat/fieldExplorer

[PRFC] Custom Field Extension Explorer

renovate[bot] committed:
Update dependency zen-observable to ^0.9.0

Signed-off-by: Renovate Bot <bot@renovateapp.com>

Commit:
Merge remote-tracking branch 'origin/master' into scaffolder-result-content

# Conflicts:
#	plugins/scaffolder/api-report.md
#	plugins/scaffolder/src/plugin.tsx

Commit:
Updated api-report.md

Signed-off-by: bnechyporenko <bnechyporenko@bol.com>

dagda1 committed:
update custom ScmAuthApi Implementation docs

Signed-off-by: Paul Cowan <paul.cowan@cutting.scot>

dagda1 committed:
run prettier

Signed-off-by: Paul Cowan <paul.cowan@cutting.scot>

Commit:
chore: Bug fixes and adding the possibility to add a default value

Signed-off-by: Mathias Bronner <mathiasb@spotify.com>
Signed-off-by: Mathias Bronner <mathiasbronner@gmail.com>

punkle committed:
move wait into try block

Signed-off-by: Brian Fletcher <brian@roadie.io>

freben committed:
Merge pull request #14565 from backstage/renovate/zen-observable-0.x

Update dependency zen-observable to ^0.9.0

Znarvl committed:
Simplified response

Co-authored-by: Johan Haals <johan@haals.se>
Signed-off-by: Simon Jakobsson <31953373+Znarvl@users.noreply.github.com>

Znarvl committed:
Removed config in changeset

Co-authored-by: Johan Haals <johan@haals.se>
Signed-off-by: Simon Jakobsson <31953373+Znarvl@users.noreply.github.com>

renovate[bot] committed:
Update dependency @azure/arm-appservice to v13.0.3

Signed-off-by: Renovate Bot <bot@renovateapp.com>

freben committed:
Merge pull request #14665 from RoadieHQ/wait-for-intialize

wait for initialize to complete before finalize

freben committed:
Merge pull request #14628 from dagda1/improve-docs-auth-custom-scmauthapi

improve the Custom ScmAuthApi Implementation section of docs/auth

Commit:
Merge remote-tracking branch 'origin/master' into cost-insight-configurable-default-currency

Rugvip committed:
integrations: add allowedInstallationOwners to config schema

Signed-off-by: Patrik Oldsberg <poldsberg@gmail.com>

Rugvip committed:
Merge pull request #14670 from backstage/dependabot/npm_and_yarn/storybook/loader-utils-1.4.2

build(deps): bump loader-utils from 1.4.1 to 1.4.2 in /storybook

freben committed:
Merge pull request #14657 from backstage/mob/versioning-apis

Update versioning policy to treat utility/service API additions as non-breaking

Rugvip committed:
docs/getting-started: clarify yarn version

Signed-off-by: Patrik Oldsberg <poldsberg@gmail.com>

Rugvip committed:
Merge pull request #14682 from backstage/rugvip/allowedowners

integrations: add allowedInstallationOwners to config schema

Rugvip committed:
Merge pull request #14683 from backstage/rugvip/classic

docs/getting-started: clarify yarn version

Znarvl committed:
removed redundant line and changed handeling etag in AWS

Signed-off-by: Simon <simme.jakobsson@gmail.com>

Znarvl committed:
Merge branch 'clean_deprecations' of github.com:Znarvl/backstage into clean_deprecations

Commit:
Incorporated the feedback

Signed-off-by: bnechyporenko <bnechyporenko@bol.com>

acierto committed:
Merge remote-tracking branch 'origin/master' into cost-insight-configurable-default-currency

acierto committed:
Merge remote-tracking branch 'origin/cost-insight-configurable-default-currency' into cost-insight-configurable-default-currency

jhaals committed:
Merge pull request #14626 from Znarvl/clean_deprecations

Cleaned up some deprecations

acierto committed:
Incorporated the feedback

Signed-off-by: bogdannechyporenko <bogdannechiporenko@gmail.com>

dagda1 committed:
Update .changeset/sour-flowers-care.md

Co-authored-by: Ben Lambert <blam@spotify.com>
Signed-off-by: Paul <paul.cowan@cutting.scot>

Commit:
chore: Fix changeset spelling and generate api reports

Signed-off-by: Mathias Bronner <mathiasbronner@gmail.com>

Commit:
chore: Fix changeset spelling again

Signed-off-by: Mathias Bronner <mathiasb@spotify.com>
Signed-off-by: Mathias Bronner <mathiasbronner@gmail.com>

acierto committed:
Merge remote-tracking branch 'origin/master' into scaffolder-result-content

freben committed:
require testing-library/prefer-screen-queries in catalog libraries

Signed-off-by: Fredrik Adelöw <freben@gmail.com>

mathiasbronner committed:
Delete nasty-pens-march.md

Signed-off-by: Mathias Bronner <mathiasbronner@gmail.com>

freben committed:
Merge pull request #14667 from dagda1/transform-errors

give the ability to supply a transformErrors function to the rjsf/core

Commit:
add GroupListPicker to the changeset text

Signed-off-by: Mathias Bronner <mathiasb@spotify.com>
Signed-off-by: Mathias Bronner <mathiasbronner@gmail.com>

vorbrodt committed:
remove paper component

Signed-off-by: Maximilian Vorbrodt <maximilian.vorbrodt@hotmail.com>

acierto committed:
Merge remote-tracking branch 'origin/master' into scaffolder-result-content

Commit:
refactor to use http-proxy-middleware

Signed-off-by: Jamie Klassen <jklassen@vmware.com>

pjungermann committed:
chore(events): apply post-merge comments

Introduces a new interface `RequestDetails` to abstract `Request`
providing access to request body and headers.

**BREAKING:** Replace `request: Request` with `request: RequestDetails` at `RequestValidator`.

**BREAKING:** Remove required field `router` at `HttpPostIngressEventPublisher.fromConfig`
  and replace it with `bind(router: Router)`.
  Additionally, the path prefix `/http` will be added inside `HttpPostIngressEventPublisher`.

Relates-to: PR #13931
Signed-off-by: Patrick Jungermann <Patrick.Jungermann@gmail.com>

freben committed:
sort out the index files

Signed-off-by: Fredrik Adelöw <freben@gmail.com>

camilaibs committed:
Merge pull request #14569 from mstergianis/docs/fix-catalogresultlistitem

docs: Replaces use of CatalogResultListItem with correct symbol

freben committed:
make a duplicate checks script

Signed-off-by: Fredrik Adelöw <freben@gmail.com>

freben committed:
Merge pull request #14696 from backstage/freben/dedup-script

make a duplicate checks script

legendik committed:
Use pluralize for catalog kind header component

Signed-off-by: Marek Šneberger <marek@sneberger.cz>

freben committed:
export as a regular express router

Signed-off-by: Fredrik Adelöw <freben@gmail.com>

legendik committed:
Add changeset

Signed-off-by: Marek Šneberger <marek@sneberger.cz>

legendik committed:
Test pluralized Technology kind

Signed-off-by: Marek Šneberger <marek@sneberger.cz>

Commit:
pass headerOptions to TemplatePage

Signed-off-by: Alex Rybchenko <arybchenko@box.com>

iamEAP committed:
Simplify base classes by using stream _construct methods

Signed-off-by: Eric Peterson <ericpeterson@spotify.com>

freben committed:
Merge pull request #14354 from backstage/freben/by-refs

added a by-refs batch endpoint for entities

freben committed:
Merge pull request #14619 from backstage/freben/by-refs-client

implement by-refs batch fetching in the catalog client

freben committed:
memoize middlewares so we don't churn resources excessively

Signed-off-by: Fredrik Adelöw <freben@gmail.com>

freben committed:
Merge pull request #14651 from backstage/freben/tests

add some eslint rules for testing-library use in tests

freben committed:
Merge pull request #14686 from backstage/freben/screen-catalog

require testing-library/prefer-screen-queries in catalog libraries

freben committed:
Use Json types from @backstage/types

Signed-off-by: Fredrik Adelöw <freben@gmail.com>

iamEAP committed:
Merge pull request #14701 from backstage/search/node-16-features

[Search] Take advantage of node 16 features

taras committed:
Fix the URL of Humanitec plugin

Signed-off-by: Taras Mankovski <taras@frontside.com>

punkle committed:
add cloudsmith plugin to backstage microsite

Signed-off-by: Brian Fletcher <brian@roadie.io>

Commit:
Merge remote-tracking branch 'origin/master' into cost-insight-configurable-default-currency

Commit:
Incorporated the feedback

Signed-off-by: bnechyporenko <bnechyporenko@bol.com>

Commit:
Incorporated the feedback

Signed-off-by: bnechyporenko <bnechyporenko@bol.com>

Commit:
Incorporated the feedback

Signed-off-by: bnechyporenko <bnechyporenko@bol.com>

freben committed:
Merge pull request #14704 from taras/patch-5

Fix the URL of Humanitec plugin

freben committed:
Merge pull request #14699 from alexrybch/scaffolder-template-header-options

Pass headerOptions to TemplatePage

freben committed:
Do not use deprecated LocationSpec

Signed-off-by: Fredrik Adelöw <freben@gmail.com>

Rugvip committed:
vale: add pattison

Signed-off-by: Patrik Oldsberg <poldsberg@gmail.com>

awanlin committed:
Removed from create-app and added to docs

Signed-off-by: Andre Wanlin <67169551+awanlin@users.noreply.github.com>

Rugvip committed:
Merge pull request #14708 from backstage/rugvip/pattison

vale: add pattison

freben committed:
Merge pull request #14707 from backstage/freben/no-deprecated-spec

Do not use deprecated LocationSpec

awanlin committed:
Changes to make Vale happy, I hope

Signed-off-by: Andre Wanlin <67169551+awanlin@users.noreply.github.com>

awanlin committed:
Removed imports

Signed-off-by: Andre Wanlin <67169551+awanlin@users.noreply.github.com>

Rugvip committed:
workflows: add workflow for posting release information on PR merge

Signed-off-by: Patrik Oldsberg <poldsberg@gmail.com>

jhaals committed:
Merge pull request #14705 from RoadieHQ/add-cloudsmith-plugin

add cloudsmith plugin to backstage microsite

jhaals committed:
Merge pull request #14703 from backstage/freben/nodep

Use Json types from @backstage/types

jhaals committed:
Merge pull request #14587 from awanlin/topic/custom-provider-annotations

Added custom provider annotation details

legendik committed:
Better changeset message

Signed-off-by: Marek Šneberger <marek@sneberger.cz>

renovate[bot] committed:
Update dependency css-loader to v6.7.2

Signed-off-by: Renovate Bot <bot@renovateapp.com>

backstage-goalie[bot] committed:
Merge pull request #14613 from backstage/renovate/css-loader-6.x-lockfile

Update dependency css-loader to v6.7.2

renovate[bot] committed:
Update dependency @rjsf/validator-ajv8 to v5.0.0-beta.13

Signed-off-by: Renovate Bot <bot@renovateapp.com>

backstage-goalie[bot] committed:
Merge pull request #14720 from backstage/renovate/rjsf-validator-ajv8-5.x-lockfile

Update dependency @rjsf/validator-ajv8 to v5.0.0-beta.13

renovate[bot] committed:
Update dependency @swc/core to v1.3.19

Signed-off-by: Renovate Bot <bot@renovateapp.com>

vorbrodt committed:
surface prop for growing or containing height of graph

Signed-off-by: Maximilian Vorbrodt <maximilian.vorbrodt@hotmail.com>

vorbrodt committed:
update changeset

Signed-off-by: Maximilian Vorbrodt <maximilian.vorbrodt@hotmail.com>

renovate[bot] committed:
Update dependency @swc/helpers to v0.4.14

Signed-off-by: Renovate Bot <bot@renovateapp.com>

vorbrodt committed:
fix spelling changeset

Signed-off-by: Maximilian Vorbrodt <maximilian.vorbrodt@hotmail.com>

backstage-goalie[bot] committed:
Merge pull request #14723 from backstage/renovate/swc-helpers-0.x-lockfile

Update dependency @swc/helpers to v0.4.14

vorbrodt committed:
update public API report of the package

Signed-off-by: Maximilian Vorbrodt <maximilian.vorbrodt@hotmail.com>

renovate[bot] committed:
Update dependency @testing-library/cypress to v8.0.7

Signed-off-by: Renovate Bot <bot@renovateapp.com>

vorbrodt committed:
Merge branch 'backstage:master' into fix-group-diagram-scroll-behaviour

backstage-goalie[bot] committed:
Merge pull request #14724 from backstage/renovate/testing-library-cypress-8.x-lockfile

Update dependency @testing-library/cypress to v8.0.7

renovate[bot] committed:
Update dependency @types/dockerode to v3.3.14

Signed-off-by: Renovate Bot <bot@renovateapp.com>

jhaals committed:
Merge pull request #14698 from legendik/bugfix/pluralizedCatalogKindHeaderComponent

Use pluralize for catalog kind header component

dagda1 committed:
update changeset description

Signed-off-by: Paul Cowan <paul.cowan@cutting.scot>

dagda1 committed:
run prettier

Signed-off-by: Paul Cowan <paul.cowan@cutting.scot>

jhaals committed:
Merge pull request #14721 from backstage/renovate/swc-core-1.x-lockfile

Update dependency @swc/core to v1.3.19

Commit:
Use promise Router & add known error in Vault backend

Signed-off-by: ivgo <ivgo@spreadgroup.com>

renovate[bot] committed:
Update dependency esbuild to v0.15.15

Signed-off-by: Renovate Bot <bot@renovateapp.com>

backstage-goalie[bot] committed:
Merge pull request #14725 from backstage/renovate/dockerode-3.x-lockfile

Update dependency @types/dockerode to v3.3.14

backstage-goalie[bot] committed:
Merge pull request #14733 from backstage/renovate/esbuild-0.x-lockfile

Update dependency esbuild to v0.15.15

freben committed:
Merge pull request #14691 from Bonial-International-GmbH/pjungermann/events/follow-up

chore(events): apply post-merge comments

freben committed:
Merge pull request #14732 from ivangonzalezacuna/fix/vault-routing

Use promise Router & add known error in Vault backend

renovate[bot] committed:
Update dependency @graphql-tools/schema to v9.0.10

Signed-off-by: Renovate Bot <bot@renovateapp.com>

renovate[bot] committed:
Update dependency eslint-plugin-react to v7.31.11

Signed-off-by: Renovate Bot <bot@renovateapp.com>

renovate[bot] committed:
Update dependency js-base64 to v3.7.3

Signed-off-by: Renovate Bot <bot@renovateapp.com>

renovate[bot] committed:
Update dependency @aws-sdk/client-sqs to v3.213.0

Signed-off-by: Renovate Bot <bot@renovateapp.com>

renovate[bot] committed:
Update dependency marked to v4.2.3

Signed-off-by: Renovate Bot <bot@renovateapp.com>

Rugvip committed:
Merge pull request #14710 from backstage/rugvip/merge-message

workflows: add workflow for posting release information on PR merge

kuangp committed:
fix(GithubMultiOrgProcessor): correctly apply team memberships

Signed-off-by: Phil Kuang <pkuang@factset.com>

jhaals committed:
Merge pull request #14346 from backstage/jhaals/lifecycle

RFC backend plugin lifecycle events

freben committed:
Merge pull request #14434 from backstage/renovate/rc-progress-3.x

Update dependency rc-progress to v3.4.1

freben committed:
Merge pull request #14679 from backstage/renovate/azure-arm-appservice-13.x-lockfile

Update dependency @azure/arm-appservice to v13.0.3

mathiasbronner committed:
chore: fix PR feedback

Signed-off-by: Mathias Bronner <mathiasbronner@gmail.com>

freben committed:
Merge pull request #14734 from backstage/renovate/eslint-plugin-react-7.x-lockfile

Update dependency eslint-plugin-react to v7.31.11

freben committed:
Merge pull request #14621 from backstage/renovate/aws-sdk-js-v3-monorepo

Update dependency @aws-sdk/client-sqs to v3.213.0

renovate[bot] committed:
Update graphqlcodegenerator monorepo

Signed-off-by: Renovate Bot <bot@renovateapp.com>

backstage-goalie[bot] committed:
Merge pull request #14494 from backstage/renovate/marked-4.x-lockfile

Update dependency marked to v4.2.3

backstage-goalie[bot] committed:
Merge pull request #14605 from backstage/renovate/graphql-tools-monorepo

Update dependency @graphql-tools/schema to v9.0.10

backstage-goalie[bot] committed:
Merge pull request #14735 from backstage/renovate/js-base64-3.x-lockfile

Update dependency js-base64 to v3.7.3

backstage-goalie[bot] committed:
Merge pull request #14738 from backstage/renovate/graphqlcodegenerator-monorepo

Update graphqlcodegenerator monorepo

renovate[bot] committed:
Update CodeMirror

Signed-off-by: Renovate Bot <bot@renovateapp.com>

renovate[bot] committed:
Update dependency @gitbeaker/browser to v35.8.0

Signed-off-by: Renovate Bot <bot@renovateapp.com>

Commit:
Incorporated the feedback

Signed-off-by: bnechyporenko <bnechyporenko@bol.com>

Commit:
Merge remote-tracking branch 'origin/master' into cost-insight-configurable-default-currency

backstage-goalie[bot] committed:
Merge pull request #14741 from backstage/renovate/codemirror

Update CodeMirror

backstage-goalie[bot] committed:
Merge pull request #14742 from backstage/renovate/gitbeaker-browser-35.x-lockfile

Update dependency @gitbeaker/browser to v35.8.0

Rugvip committed:
Merge pull request #14403 from vorbrodt/fix-group-diagram-scroll-behaviour

Fix group diagram scroll behaviour

Rugvip committed:
Merge pull request #14386 from awanlin/topic/kind-sidebarsubmenu

Added SidebarSubmenu for all Kinds

renovate[bot] committed:
Update dependency @gitbeaker/node to v35.8.0

Signed-off-by: Renovate Bot <bot@renovateapp.com>

renovate[bot] committed:
Update dependency @google-cloud/container to v4.4.0

Signed-off-by: Renovate Bot <bot@renovateapp.com>

backstage-goalie[bot] committed:
Merge pull request #14745 from backstage/renovate/gitbeaker-node-35.x-lockfile

Update dependency @gitbeaker/node to v35.8.0

backstage-goalie[bot] committed:
Merge pull request #14746 from backstage/renovate/google-cloud-container-4.x-lockfile

Update dependency @google-cloud/container to v4.4.0

renovate[bot] committed:
Update dependency @tanstack/react-query to v4.16.1

Signed-off-by: Renovate Bot <bot@renovateapp.com>

renovate[bot] committed:
Update dependency @uiw/react-codemirror to v4.15.1

Signed-off-by: Renovate Bot <bot@renovateapp.com>

backstage-goalie[bot] committed:
Merge pull request #14747 from backstage/renovate/tanstack-query-monorepo

Update dependency @tanstack/react-query to v4.16.1

backstage-goalie[bot] committed:
Merge pull request #14748 from backstage/renovate/uiw-react-codemirror-4.x-lockfile

Update dependency @uiw/react-codemirror to v4.15.1

renovate[bot] committed:
Update dependency concurrently to v7.6.0

Signed-off-by: Renovate Bot <bot@renovateapp.com>

acierto committed:
Merge remote-tracking branch 'origin/master' into scaffolder-result-content

# Conflicts:
#	plugins/scaffolder/src/components/TemplatePage/TemplatePage.tsx

backstage-goalie[bot] committed:
Merge pull request #14750 from backstage/renovate/concurrently-7.x-lockfile

Update dependency concurrently to v7.6.0

renovate[bot] committed:
Update dependency @aws-sdk/client-sqs to v3.215.0

Signed-off-by: Renovate Bot <bot@renovateapp.com>

backstage-goalie[bot] committed:
Merge pull request #14751 from backstage/renovate/aws-sdk-js-v3-monorepo

Update dependency @aws-sdk/client-sqs to v3.215.0

renovate[bot] committed:
Update dependency elastic-builder to v2.17.3

Signed-off-by: Renovate Bot <bot@renovateapp.com>

renovate[bot] committed:
Update dependency eslint to v8.28.0

Signed-off-by: Renovate Bot <bot@renovateapp.com>

backstage-goalie[bot] committed:
Merge pull request #14755 from backstage/renovate/elastic-builder-2.x-lockfile

Update dependency elastic-builder to v2.17.3

backstage-goalie[bot] committed:
Merge pull request #14756 from backstage/renovate/eslint-8.x-lockfile

Update dependency eslint to v8.28.0

renovate[bot] committed:
Update dependency msw to ^0.49.0

Signed-off-by: Renovate Bot <bot@renovateapp.com>

renovate[bot] committed:
Update dependency typescript to v4.9.3

Signed-off-by: Renovate Bot <bot@renovateapp.com>

jhaals committed:
Merge pull request #14678 from mathiasbronner/master

chore: GroupListPicker - Bug fixes and adding the possibility to add a default value

freben committed:
Merge pull request #14736 from kuangp/fix/GithubMultiOrgProcessor

fix(GithubMultiOrgProcessor): correctly apply team memberships

jhaals committed:
Merge pull request #14634 from acierto/cost-insight-configurable-default-currency

Cost insight configurable default currency

Rugvip committed:
workflows/automate_merge_message: fix for forks

Signed-off-by: Patrik Oldsberg <poldsberg@gmail.com>

jhaals committed:
Merge pull request #14766 from backstage/rugvip/target

workflows/automate_merge_message: fix for forks

benjdlambert committed:
chore: fixing types for `uiSchema`
Signed-off-by: blam <ben@blam.sh>

benjdlambert committed:
chore: fixing api-reports

Signed-off-by: blam <ben@blam.sh>

jhaals committed:
Merge pull request #14764 from backstage/renovate/typescript-4.x-lockfile

Update dependency typescript to v4.9.3

camilaibs committed:
fix(search-backend-node): max listeners exceeded warning

Signed-off-by: Camila Belo <camilaibs@gmail.com>

jhaals committed:
Merge pull request #14761 from backstage/renovate/msw-0.x

Update dependency msw to ^0.49.0

benjdlambert committed:
Merge pull request #14132 from acierto/scaffolder-result-content

Make scaffolder last step (summary overview) configurable by the user

Rugvip committed:
workflows/automate_merge_message: fix diff point

Signed-off-by: Patrik Oldsberg <poldsberg@gmail.com>

camilaibs committed:
apply review suggestions

Signed-off-by: Camila Belo <camilaibs@gmail.com>

Rugvip committed:
Merge pull request #14774 from backstage/rugvip/diffail

workflows/automate_merge_message: fix diff point

benjdlambert committed:
Merge pull request #14660 from dagda1/stepper-form-data

initialize all formData in next/Stepper

iamEAP committed:
Merge pull request #14772 from backstage/search-backend-node/fix-max-listeners-exceeded-warning

[Search] Fix max listeners exceeded warning

Rugvip committed:
Merge pull request #14641 from backstage/emmaindal/search-alpha-label

[Search] remove search alpha label in header

freben committed:
use the right msw version

Signed-off-by: Fredrik Adelöw <freben@gmail.com>
